Fb shared component test 7220

labkey-chrisj opened the pull request:
#### Rationale
This provides support for testing components via the core-Components view, and creates tests specificially for grid
pulls the Pager component out of Biologics and adapts it for common use, updates QueryGrid to use it for paging

#### Related Pull Requests
https://github.com/LabKey/biologics/pull/652

#### Changes
Creates a page class that wraps the core-components.view action
Creates forking logic within QueryGrid test component to support transition from old QueryGridPanel toward new QueryGrid

labkey-chrisj committed:
Create Page class for CoreComponentsTest action, and a test class for QueryGrid testing
Refactor Pager from Biologics and use it as GridBar subcomponent
Enable DropdownButtonGroup to get its button's text

labkey-chrisj committed:
Restore support for biologicsReportTest

labkey-chrisj committed:
relax subsequent-state-check to allow the component to be used in cases where checking it results in indeterminate state

labkey-chrisj committed:
Provide method to get QueryGridPanel component until the transition to QueryGrid is complete

labkey-chrisj committed:
Make Omnibox aware of its opened/closed state; enable it to close itself

labkey-chrisj committed:
Add tests to cover grid row selection with single-page and mulit-page filter selections

labkey-chrisj committed:
QueryGrid component updates to prepare for a smooth transition away from QueryGridPanel to new QueryGrid

labkey-chrisj commented on src/org/labkey/test/components/glassLibrary/grids/GridBar.java:
```diff
@@ -28,12 +27,14 @@
 public class GridBar extends WebDriverComponent<GridBar.ElementCache>
 {
     final private WebElement _gridBarElement;
+    final private WebElement _containerElement;
```
Just a general note on this 'containerElement'- because the newer QueryGrid component moves things like the OmniBox and selection info controls out of the _gridBarElement, we may want to consider whether to expand that element to contain them, or re-think whether or not GridBar ought to be replaced

labkey-chrisj committed:
add 'summary' text to allow test to assert/check expected range numbers

labkey-chrisj committed:
if clear-all button is not present, use checkbox to deselect all

labkey-chrisj committed:
Add test coverage for issue 41161

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into fb_sharedComponentTest_7220

labkey-chrisj committed:
enable Omnibox test component to support multi-word columns and operator expressions

labkey-chrisj committed:
coverage for multi-word filter operators

labkey-chrisj committed:
add coverage for scenarios with filters and selections

labkey-tchad commented on src/org/labkey/test/components/glassLibrary/components/OmniBox.java:
```diff
@@ -144,20 +166,42 @@ private void sendClearValue()
-        WebDriverWrapper.waitFor(()->  getValue(expectedValue) != null,
-                "Expect a new value to be present with [" + expectedValue + "]", 1500);
+        WebDriverWrapper.waitFor(()->  getValue(expectedFilterText.toString()) != null,
+                "Expect a new value to be present with [" + expectedFilterText.toString() + "]", 1500);
```
Log the retry so that somebody could investigate whether it is having to do so too frequently.

labkey-chrisj replied:
done, not sure why github doesn't mark this as outdated

labkey-tchad commented on src/org/labkey/test/components/glassLibrary/grids/GridBar.java:
```diff
@@ -68,11 +69,46 @@ public File exportData(ExportType exportType)
         return getWrapper().doAndWaitForDownload(()->exportButton.click());
     }
 
+    /**
+     * A word of caution for callers; the Pager is not aware of the grid state.
+     * To navigate (next page, previous, first/last) use methods on GridBar to do it-
+     * they will synchronize against grid updates
+     * @return
+     */
+    public Pager pager()
+    {
+        return new Pager.PagerFinder(getDriver()).waitFor(this);
+    }
```
Why not have the `Pager` handle synchronization? Pass the `_responsiveGrid` in its constructor and have the `Pager` make the `doAndWaitForUpdate` calls to the grid.

labkey-chrisj replied:
There are cases where the pager interacts with not-a-grid (specifically, in Biologics, the reportList).
I initially did that and passed in the _responsiveGrid to synchronize internally, but opted to make it relatively dumb and have the methods calling it (from QueryGrid) handle the synchronizing.
I'm willing to pass in the grid and locate the synchronization inside this component, but then I'd have to fix the BiologicsReportTest to not use it.  The upside to doing that would mean only having to support grid usage, and that would simplify a lot of code

labkey-chrisj replied:
...later in this review you suggested using an interface to deal with the coupling issues of wanting to pass in the grid.  I like that as a concept and am going to make a brave attempt, but may need help implementing that.

labkey-chrisj replied:
Actually, I did it and it works

labkey-tchad commented on src/org/labkey/test/components/glassLibrary/grids/GridBar.java:
```diff
@@ -83,51 +119,25 @@ public int getRecordCount()
-
+        _responsiveGrid.doAndWaitForUpdate(()-> pager().clickNextButton());
         return _responsiveGrid;
     }
 
     // Should this really return a grid object?
     public ResponsiveGrid getPreviousPage()
     {
-        WebElement prevButton = Locators.pgLeftButton.waitForElement(this, 2500);
-        if (prevButton.isEnabled())
-        {
-            _responsiveGrid.doAndWaitForUpdate(()->prevButton.click());
-        }
+        _responsiveGrid.doAndWaitForUpdate(()-> pager().clickPrevious());
         return _responsiveGrid;
     }
```
`get*` implies a non-mutating operation. I'd go with `clickNext`/`clickPrevious` to match the `Pager`

labkey-chrisj replied:
Good call.  I'd been trying not to mess with established naming unless necessary but I like it.

labkey-chrisj replied:
done

labkey-tchad commented on src/org/labkey/test/components/glassLibrary/grids/Pager.java:
```diff
@@ -0,0 +1,194 @@
+{
+    private final WebElement _el;
+    private final WebDriver _driver;
+    //private final ResponsiveGrid _grid;
```
I presume the `ResponsiveGrid` is omitted so that the biologics report list can use the pager. You could make a simple interface for components that can update:
```
public interface UpdatingComponent
{
    public static final UpdatingComponent DEFAULT = Runnable::run;

    void doAndWaitForUpdate(Runnable func);
}
```
`Pager` would take an `UpdatingComponent` in its constructor. Classes would either implement it and pass themselves to the `Pager` or just pass `UpdatingComponent.DEFAULT` to it. Alternately, you could leave the existing `Pager` constructor signature and have it pass through `UpdatingComponent.DEFAULT`.

labkey-chrisj replied:
Good guess.  Thanks for providing enough guidance here to help me create the interface and wire it all up

labkey-tchad commented on src/org/labkey/test/components/glassLibrary/grids/Pager.java:
```diff
@@ -0,0 +1,194 @@
+        return !elementCache().prevButton.getAttribute("disabled").equals("true");
+    }
+
+    public Pager clickNextButton()
```
Use consistent method names
```suggestion
    public Pager clickNext()
```

labkey-chrisj replied:
done

labkey-tchad commented on src/org/labkey/test/components/glassLibrary/grids/Pager.java:
```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright (c) 2019 LabKey Corporation. All rights reserved. No portion of this work may be reproduced in
+ * any form or by any electronic or mechanical means without written permission from LabKey Corporation.
+ */
```
Remove inappropriate copyright.

labkey-tchad commented on src/org/labkey/test/components/glassLibrary/grids/Pager.java:
```diff
@@ -0,0 +1,194 @@
+ * Copyright (c) 2019 LabKey Corporation. All rights reserved. No portion of this work may be reproduced in
+ * any form or by any electronic or mechanical means without written permission from LabKey Corporation.
+ */
+package org.labkey.test.components.glassLibrary.grids;
```
Don't use the `glassLibrary` package. It'll just be one more thing to move once we decide to refactor. I've been using `org.labkey.test.components.labkey.ui`.
```suggestion
package org.labkey.test.components.labkey.ui.grids;
```

labkey-chrisj replied:
I've been putting off making a new package/migrating everything into it, largely because I don't have a clear vision for what the new destination should look like and I didn't want to make the changes and then be told I did it wrong.  I'll just do my best for now.

labkey-tchad commented on src/org/labkey/test/components/glassLibrary/grids/Pager.java:
```diff
@@ -0,0 +1,194 @@
+            String value = elementCache().counts.getAttribute("data-total");
+            return Integer.parseInt(value);
+        }
+    }
```
I suggest combining `start()`, `end()`, and `total()` into a single `getPaginationInfo()` method that gathers all three and wraps them in a little class like:
```
public static class PaginationInfo
{
    public final int _start, _end, _total;
    PaginationInfo(int start, int end, int total)
    {
        _start = start;
        _end = end;
        _start = total;
    }
}
```

labkey-chrisj replied:
I confess; I'm not 100% sure what that will get us.  I agree that the current implementation is pretty verbose and would like to clean it up.  Let's talk about how this can be used

labkey-chrisj replied:
Right now, ` summary()` (line 138) consumes the text from the 'counts' element, which only exists in the new GridPanel.

labkey-tchad commented on src/org/labkey/test/components/react/ReactCheckBox.java:
```diff
@@ -78,6 +77,10 @@ public void set(@NotNull Boolean checked)
+        // Sometimes (like when the box has a ternary state, as the 'select all' box in grids)
+        // checking the box won't result in a checked state- it will instead land on 'indeterminate'.
+        WebDriverWrapper.waitFor(() -> checked == isChecked() || checked == isIndeterminate(),
+                "Failed to " + (checked ? "check" : "uncheck") + " checkbox", 1000);
```
As discussed. Create a `set` overload that takes an expected state.

labkey-chrisj replied:
done

labkey-tchad commented on src/org/labkey/test/tests/component/QueryGridTest.java:
```diff
@@ -0,0 +1,443 @@
+        _containerHelper.createProject(getProjectName(), null);
+    }
+
+    @Before
+    public void preTest() throws Exception
+    {
+        goToProjectHome();
+    }
+
```
This seems unnecessary (I should probably remove it from the template).
All of the tests start with a `CoreComponentsTestPage.beginAt`

labkey-tchad commented on src/org/labkey/test/tests/component/QueryGridTest.java:
```diff
@@ -0,0 +1,443 @@
+        sampleSetDataGenerator.deleteDomain(createDefaultConnection());
+    }
+
+    @Ignore // until https://www.labkey.org/home/Developer/issues/issues-details.view?issueId=41169 is resolved
+    @Test
+    public void testEmptyNotEmptyFilter() throws Exception
```
Why `@Ignore` this test? It looks like the "has any value" check will pass with the current (incorrect) behavior which seems like the right approach.

labkey-chrisj replied:
If it was behaving correctly, it would say 1- 20 of 30 but it instead current behavior has it saying 1-20 of 35 because the custom rows (which don't have values in Int Column) aren't filtered out.  
`@Ignore` is there so it's not noisy until the bug is addressed.  If that's not right, I'm happy to change it

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into fb_sharedComponentTest_7220

labkey-chrisj committed:
create UpdatingComponent interface for use with paging components

labkey-chrisj committed:
Because sometimes a checkbox will correctly land in an 'indeterminate' state when clicked from an un-cheded state,
this change allows the component (and ones that call it) to decide what intended state to verify

labkey-chrisj committed:
cr feedback
- make Pager take UpdatingComponent and have grid implement it
- move Pager class out of glassLibrary package location
- update some public method names and refactor usages

labkey-chrisj committed:
check filter list for null before attempting to reference it

labkey-chrisj committed:
Move component files out of glasslibrary package

labkey-chrisj committed:
Revert "Move component files out of glasslibrary package"

This reverts commit 43526df3

labkey-chrisj committed:
put breaks in switch statement to avoid cascading through case handlers

labkey-chrisj committed:
add backwards-support for QueryGridPanel component - checks for whether next and prev buttons are disabled

labkey-chrisj committed:
handle situations where next and prev buttons aren't present

labkey-chrisj committed:
put QueryGridTest into dailyB until we have a component-test suite

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into fb_sharedComponentTest_7220

# Conflicts:
#	src/org/labkey/test/util/TestDataGenerator.java

labkey-chrisj committed:
remove @Ignore attribute on test, comment part that will fail until issue 41169 is fixed

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into fb_sharedComponentTest_7220
